### ISSUE TITLE
Bonuses can only be used once per round (phase 10)

### ISSUE THREAD

**Issue by tsjensen** [tsjensen/civbuddy#1]
Currently, CivBuddy allows the player to use bonuses from civilization cards for more than one purchase per round.
Rule 43.10 of the original game explicitly states that bonuses from one civilization card may be used only for one purchase per round. If two bonuses are available from one civilization card, only one of them may be used.

**Comment by tsjensen** [comment]
@toadwana Are you sure? After careful study of rule 43.10, it seems to me that using a bonus more than once is allowed (in fact, it is done with _Architecture_ in the example). Instead, only one of the card's bonuses may be used per purchase (#3).

**Comment by tsjensen** [comment]
@asheiduk Opinion on this?

**Comment by asheiduk** [comment]
Quoting the german rules from 43.10 (emphasis mine):

> Die Gutschriften können in beliebiger Weise kombiniert werden, wobei jede _Karte_ pro _Kauf_ nur einmal verwendet werden darf. 

I read _Kauf_ as "round" here. Otherwise that text would define a "rule" which would have no effect at all. Considering that text as an explanation is also quite stretched.

Also the complete game acts that in one phase one player makes "atomic" actions i.e. without any further subdivisions of the phase.

> Eine _Gutschrift_ kann erst in der darauffolgenden Runde wieder für den Kauf einer anderen Karte verwendet werden.

IMO the Architecture example stresses only the point, that only one bonus can be used and not the sum of both.

TL;DR: Each bonus-card can be used only once per round and only one bonus of the card can be used in that round.

**Comment by tsjensen** [comment]
Agreed. I would say you basically choose some civilization cards already in your posession and *assign* them to a purchased card. Each assignment is final for the remainder of the round. Ok?

**Comment by tsjensen** [comment]
From the English rules to the original game (emphasis mine):
> **17.11** The same credit from a civilization card already owned *can be applied toward different cards in the same turn*, but it cannot be applied more than once toward the same card.
For example, Player B has Mysticism and wishes to acquire two sciences, Astronomy and Coinage. Mysticism's 20 point science credit can be applied once to the acquisition of each for a total credit of 40 points.

Admittedly, the German rules are not that precise. CivBuddy v2 will have a rule option to choose the behavior.

**Comment by asheiduk** [comment]
I think the most important thing is: The German rules differ from the English rules!

**Comment by tsjensen** [comment]
Right!
Frankly, I was quite surprised to realize that. And that's part of the reason why we will have rule options. 😏

**Comment by tsjensen** [comment]
@asheiduk I just started thinking about how to implement this. It seems quite complicated. How did you do it?

I assume that you basically have to "assign" owned cards to newly acquired cards in order to activate their bonuses. Obviously, each card could only be assigned to one newly acquired card.

Consider this use case:

1. **Have** *Mysticism*
1. **Plan** *Music* (for 55 instead of 60)
  This must update the cost of *Medicine* from 120 to 125 (and many other cards, too)
1. **Plan** *Medicine* (for 125 instead of 140)
1. **Buy** *Medicine* and *Music* for 180 total, using the *Mysticism* bonus for *Medicine* (as it's 20, which is better than the 5 for Music)

So, whenever a card is planned or unplanned, a lot of "current cost" values change, along with their blue credit bars.

**Comment by asheiduk** [comment]
Well it's complex, not complicated :-)

My code assumes, that I have a set of cards "at hand"  and a second *set* of cards I want. First I assign each of the "wanted" cards appropriate bonus cards eventually more than required. I call that "overpay". In a second run I optimize that: If moving an "at hand" card to another "wanted" card improves utilisation of that bonus card then also the total cost of all "wanted" cards is reduced. Of course the previous owner of the bonus card might draw another still unused "at hand" bonus card.

Applied to your example:

- "at hand" set: " *Mysticism*
- "wanted set": *Music*, *Medicine*

1) Assign *Mysticism* to *Music* -> total cost: 55 + 140 = 195
1) Try to move *Mysticism* to *Medicine*: *Mysticism* not costs 60, *Medicine* 120, total: 180
Total cost is lower, so keep that.

Exactly here is a small difference to your train of though: At no stage I'm thinking, that *Medicine* costs 125 because *Mysticism* still "holds" 5 points. Either a bonus card is assigned completely to another card or not.

Wrapping around my head around that stuff after so long is quite difficult.

**Comment by tsjensen** [comment]
It's not even very complex, but it's more complicated than expected ... 👨‍🔬 😄 

Thanks for the insight! Do you display the current cost of cards? If so, then what cost do you show for *Medicine* when the user has already planned to buy *Music*?
We want to be able to tell if the user can afford a purchase, so the amounts shown must be the amounts that will be due.

**Comment by asheiduk** [comment]
I must recede in one point from my previous comment:

When my user marked *Mysticism* as "on hand" then
- *Medicine* is 140 - 20 = 120
- *Music* is 60 - 5 = 55

When the user now marks *Musik* as "want" then
- *Medicine* is 140 - 15 = 125
- *Music* is 60 - 5 = 55

Here 125 is displayed for *Medicine*. That step eluded me because I immediately continued with:

When the user now marks *Medicine* as "want also" then
- *Medicine* is 140 - 20 = 120
- *Music* is 60 - 0 = 60

**Comment by tsjensen** [comment]
Released.
